python/ray/llm/_internal/batch/stages/vllm_engine_stage.py

```diff
@@ -0,0 +1,530 @@
+        self.max_pending_requests = max_pending_requests or math.ceil(
+            self.engine_kwargs["max_num_seqs"] * pp_size * 1.1
+        )
+        if self.max_pending_requests > 0:
+            logger.info("Max pending requests is set to %d", self.max_pending_requests)
+
+        # Create an LLM engine.
+        self.llm = vLLMEngineWrapper(
+            model=self.model,
+            idx_in_batch_column=self.IDX_IN_BATCH_COLUMN,
+            disable_log_stats=False,
+            max_pending_requests=self.max_pending_requests,
+            runtime_env=self.runtime_env,
+            **self.engine_kwargs,
+        )
+
+    def normalize_engine_kwargs(
+        self,
+        task_type: str,
+        engine_kwargs: Dict[str, Any],
+    ) -> Dict[str, Any]:
+        """
+        Normalize the engine kwargs.
+
+        Args:
+            task_type: The task to use for the vLLM engine (e.g., "generate", "embed", etc).
+            engine_kwargs: The kwargs to normalize.
+
+        Returns:
+            The normalized kwargs.
+        """
+        # Remove model from engine kwargs if set.
+        model = engine_kwargs.pop("model", None)
+        if model is not None and model != self.model:
+            logger.warning(
+                "The model set in engine kwargs (%s) is different from the "
+                "stage (%s). Please remove 'model' from engine kwargs.",
+                model,
+                self.model,
+            )
+
+        # Override the task if it is different from the stage.
+        task = engine_kwargs.get("task", task_type)
+        if task != task_type:
+            logger.warning(
+                "The task set in engine kwargs (%s) is different from the "
+                "stage (%s). Overriding the task in engine kwargs to %s.",
+                task,
+                task_type,
+                task_type,
+            )
+        engine_kwargs["task"] = task_type
+
+        # Override vLLM default configs. Note that this is only effective
+        # when the config is not set by users.
+        engine_kwargs.setdefault("gpu_memory_utilization", 0.95)
+        engine_kwargs.setdefault("use_v2_block_manager", True)
+        engine_kwargs.setdefault("enable_prefix_caching", False)
+        engine_kwargs.setdefault("enforce_eager", False)
+        engine_kwargs.setdefault("pipeline_parallel_size", 1)
+        engine_kwargs.setdefault("max_num_seqs", 256)
+        engine_kwargs.setdefault("tensor_parallel_size", 1)
+        engine_kwargs.setdefault("max_logprobs", 0)
+
+        # FlashInfer does not support bfloat16 activations, so we enforce float16
+        # dtype in this case.
+        env_vars = self.runtime_env.get("env_vars", {})
+        if (attn_backend := env_vars.get("VLLM_ATTENTION_BACKEND", None)) is not None:
+            if (
+                attn_backend == "FLASHINFER"
+                and engine_kwargs.get("kv_cache_dtype", "auto") == "fp8"
```

Comment:
shouldn't we just handle this condition the engine and error out instead of magically fixing the config?

Reply:
Ditto. We can remove everything magical and leverage vLLM.